File: 09-Recursion/cpp/06-permutation-sequence.cpp
class Solution {
public:
    string getPermutation(int n, int k) {
        string res = "";
        vector<int> nums;
        vector<int> fact(n + 1, 1);

        for (int i = 1; i <= n; ++i)
            nums.push_back(i);

        for (int i = 2; i <= n; ++i)
            fact[i] = fact[i - 1] * i;

        --k;

        for (int i = n - 1; i >= 0; --i) {
            int j = k / fact[i];
            k %= fact[i];
            res += to_string(nums[j]);
            nums.erase(nums.begin() + j);
        }

        return res;
    }
};

File: 09-Recursion/python/06_permutation_sequence.py
class Solution:
    def getPermutation(self, n: int, k: int) -> str:
        nums = list(range(1, n + 1))
        fact = [1] * (n + 1)

        for i in range(2, n + 1):
            fact[i] = fact[i - 1] * i

        k -= 1
        res = []

        for i in range(n - 1, -1, -1):
            j = k // fact[i]
            k %= fact[i]
            res.append(str(nums[j]))
            nums.pop(j)

        return "".join(res)
